pkg/openfeature/api.go

```diff
@@ -0,0 +1,119 @@
+package openfeature
+
+import (
+	"errors"
+	"github.com/go-logr/logr"
+	"github.com/open-feature/go-sdk/pkg/openfeature/internal"
+	"sync"
+)
+
+// evaluationAPI wraps OpenFeature evaluation API functionalities
+type evaluationAPI struct {
+	defaultProvider FeatureProvider
+	namedProviders  map[string]FeatureProvider
+	hks             []Hook
+	evalCtx         EvaluationContext
+	logger          logr.Logger
+	mu              sync.RWMutex
+}
+
+// newEvaluationAPI is a helper to generate an API. Used internally
+func newEvaluationAPI() evaluationAPI {
+	return evaluationAPI{
+		defaultProvider: NoopProvider{},
+		namedProviders:  map[string]FeatureProvider{},
+		hks:             []Hook{},
+		evalCtx:         EvaluationContext{},
+		logger:          logr.New(internal.Logger{}),
+		mu:              sync.RWMutex{},
+	}
+}
+
+// setProvider sets the default provider of the evaluationAPI. Returns an error if FeatureProvider is nil
+func (api *evaluationAPI) setProvider(provider FeatureProvider) error {
+	api.mu.RLock()
+	defer api.mu.RUnlock()
+
+	if provider == nil {
+		return errors.New("default provider cannot be set to nil")
+	}
+
+	api.defaultProvider = provider
+	api.logger.V(internal.Info).Info("set global provider", "name", provider.Metadata().Name)
+
+	return nil
+}
+
+// setProvider sets a provider with client name. Returns an error if FeatureProvider is nil
+func (api *evaluationAPI) setNamedProvider(clientName string, provider FeatureProvider) error {
+	api.mu.RLock()
+	defer api.mu.RUnlock()
+
+	if provider == nil {
+		return errors.New("provider cannot be set to nil")
+	}
+
+	api.namedProviders[clientName] = provider
+	api.logger.V(internal.Info).Info("set named provider provider", "name", "providerName", clientName, provider.Metadata().Name)
+
+	return nil
+}
+
+func (api *evaluationAPI) provider() FeatureProvider {
+	api.mu.RLock()
+	defer api.mu.RUnlock()
+
+	return api.defaultProvider
+}
+
+func (api *evaluationAPI) setEvaluationContext(evalCtx EvaluationContext) {
+	api.mu.RLock()
```

Comment:
I see that in many setters the lock is only in read-mode. Please, update them with a call to `Lock()` to lock in write-mode

Reply:
Updated with the latest commit. thanks for noticing this